[xexpression]`xshared_expression` cannot take in `xexpressions` without strides and other attributes

Reproducing from a cpp file:

```
#include <iostream>
#include <xtensor/xarray.hpp>
#include <xtensor/xio.hpp>
#include <xtensor/xview.hpp>
#include <xtensor/xexpression.hpp>

using int8 = std::int8_t;

template <typename T>
void make_shared(T&& expr){
    auto expression = xt::make_xshared(std::move(expr));
}

int main(){
    xt::xarray<double> arr1 {1.0, 2.0, 3.0};
    xt::xarray<double> arr2 {5.0, 6.0, 7.0};

    make_shared(arr1); // works
    make_shared(std::move(arr1)); // also works
    make_shared(xt::eval(arr1+arr2)); // also works
    make_shared(arr1+arr2); // fails
}
```

The error seems to be that `xshared_expression` requires a `strides_type`, and general `xfunctions` do not have this attribute:

```
(idp3) ❯ g++ -std=c++17 -I/Users/ray_zhang/anaconda3/envs/idp3/include hmm.cpp
In file included from hmm.cpp:2:
In file included from /Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xarray.hpp:19:
In file included from /Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xcontainer.hpp:23:
In file included from /Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xmath.hpp:23:
In file included from /Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xaccumulator.hpp:17:
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xexpression.hpp:371:42: error: no type named 'strides_type' in 'xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
        using strides_type = typename E::strides_type;
                             ~~~~~~~~~~~~^~~~~~~~~~~~
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xexpression.hpp:503:16: note: in instantiation of template class 'xt::xshared_expression<xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> >'
      requested here
        return xshared_expression<E>(std::make_shared<E>(expr));
               ^
hmm.cpp:11:27: note: in instantiation of function template specialization 'xt::make_xshared<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> >' requested here
    auto expression = xt::make_xshared(std::move(expr));
                          ^
hmm.cpp:20:5: note: in instantiation of function template specialization 'make_shared<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> >' requested here
    make_shared(arr1+arr2); // fails
    ^
In file included from hmm.cpp:2:
In file included from /Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xarray.hpp:19:
In file included from /Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xcontainer.hpp:23:
In file included from /Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xmath.hpp:23:
In file included from /Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xaccumulator.hpp:17:
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xexpression.hpp:372:46: error: no type named 'backstrides_type' in 'xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
        using backstrides_type = typename E::backstrides_type;
                                 ~~~~~~~~~~~~^~~~~~~~~~~~~~~~
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xexpression.hpp:375:48: error: no type named 'inner_strides_type' in 'xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
        using inner_strides_type = typename E::inner_strides_type;
                                   ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xexpression.hpp:376:52: error: no type named 'inner_backstrides_type' in 'xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
        using inner_backstrides_type = typename E::inner_backstrides_type;
                                       ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xexpression.hpp:409:32: error: no member named 'strides' in 'xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
        XTENSOR_FORWARD_METHOD(strides);
        ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xexpression.hpp:324:39: note: expanded from macro 'XTENSOR_FORWARD_METHOD'
        -> decltype(std::declval<E>().name())  \
                    ~~~~~~~~~~~~~~~~~ ^
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xexpression.hpp:411:32: error: no member named 'data' in 'xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
        XTENSOR_FORWARD_METHOD(data);
        ~~~~~~~~~~~~~~~~~~~~~~~^~~~~
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xexpression.hpp:324:39: note: expanded from macro 'XTENSOR_FORWARD_METHOD'
        -> decltype(std::declval<E>().name())  \
                    ~~~~~~~~~~~~~~~~~ ^
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xexpression.hpp:413:32: error: no member named 'data_offset' in 'xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
        XTENSOR_FORWARD_METHOD(data_offset);
        ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xexpression.hpp:324:39: note: expanded from macro 'XTENSOR_FORWARD_METHOD'
        -> decltype(std::declval<E>().name())  \
                    ~~~~~~~~~~~~~~~~~ ^
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xexpression.hpp:415:32: error: no member named 'storage' in 'xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
        XTENSOR_FORWARD_METHOD(storage);
        ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xexpression.hpp:324:39: note: expanded from macro 'XTENSOR_FORWARD_METHOD'
        -> decltype(std::declval<E>().name())  \
                    ~~~~~~~~~~~~~~~~~ ^
In file included from hmm.cpp:1:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/iostream:38:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/ios:216:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__locale:15:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string:470:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string_view:169:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__string:56:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:643:
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:2057:9: error: no matching constructor for initialization of 'xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
      : __value_(_VSTD::forward<_Args>(_VSTD::get<_Indexes>(__args))...) {}
        ^        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:2159:9: note: in instantiation of function template specialization 'std::__1::__compressed_pair_elem<xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>, 1,
      false>::__compressed_pair_elem<xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > &, 0>' requested here
        _Base2(__pc, _VSTD::move(__second_args),
        ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:3572:16: note: in instantiation of function template specialization
      'std::__1::__compressed_pair<std::__1::allocator<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> >, xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>
      >::__compressed_pair<std::__1::allocator<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > &, xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > &>'
      requested here
            :  __data_(piecewise_construct, _VSTD::forward_as_tuple(__a),
               ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:4230:26: note: in instantiation of function template specialization 'std::__1::__shared_ptr_emplace<xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>,
      std::__1::allocator<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > >::__shared_ptr_emplace<xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > &>' requested here
    ::new(__hold2.get()) _CntrlBlk(__a2, _VSTD::forward<_Args>(__args)...);
                         ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:4600:29: note: in instantiation of function template specialization 'std::__1::shared_ptr<xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>
      >::make_shared<xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long,
      4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long,
      4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > &>' requested here
    return shared_ptr<_Tp>::make_shared(_VSTD::forward<_Args>(__args)...);
                            ^
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xexpression.hpp:503:43: note: in instantiation of function template specialization 'std::__1::make_shared<xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>,
      xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > &>' requested here
        return xshared_expression<E>(std::make_shared<E>(expr));
                                          ^
hmm.cpp:20:5: note: in instantiation of function template specialization 'make_shared<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> >' requested here
    make_shared(arr1+arr2); // fails
    ^
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xfunction.hpp:557:9: note: candidate constructor not viable: no known conversion from 'xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> >' to
      'const xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>' for 1st argument
        xfunction(const xfunction&) = default;
        ^
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xfunction.hpp:560:9: note: candidate constructor not viable: no known conversion from 'xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> >' to
      'xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>' for 1st argument
        xfunction(xfunction&&) = default;
        ^
/Users/ray_zhang/anaconda3/envs/idp3/include/xtensor/xfunction.hpp:553:9: note: candidate template ignored: requirement '!std::is_base_of<std::decay_t<xexpression<xfunction<plus<double>, double, const
      xarray_container<uvector<double, allocator<double> >, xt::layout_type::row_major, svector<unsigned long, 4, allocator<unsigned long>, true>, xtensor_expression_tag> &, const xarray_container<uvector<double,
      allocator<double> >, xt::layout_type::row_major, svector<unsigned long, 4, allocator<unsigned long>, true>, xtensor_expression_tag> &> > &>, self_type>::value' was not satisfied [with Func =
      xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > &, CTA = <>]
        xfunction(Func&& f, CTA&&... e) noexcept;
        ^
9 errors generated.
```

Here is what happens if we remove the declarations:

```
(cling) ❯ g++ -std=c++17 -I/Users/ray_zhang/anaconda3/envs/cling/include hmm.cpp
In file included from hmm.cpp:2:
In file included from /Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xarray.hpp:19:
In file included from /Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xcontainer.hpp:23:
In file included from /Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xmath.hpp:23:
In file included from /Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xaccumulator.hpp:17:
/Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xexpression.hpp:409:32: error: no member named 'strides' in
      'xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long,
      4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
        XTENSOR_FORWARD_METHOD(strides);
        ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
/Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xexpression.hpp:324:39: note: expanded from macro 'XTENSOR_FORWARD_METHOD'
        -> decltype(std::declval<E>().name())  \
                    ~~~~~~~~~~~~~~~~~ ^
/Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xexpression.hpp:503:16: note: in instantiation of template class
      'xt::xshared_expression<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>
      >' requested here
        return xshared_expression<E>(std::make_shared<E>(expr));
               ^
hmm.cpp:11:27: note: in instantiation of function template specialization 'xt::make_xshared<xt::xfunction<xt::detail::plus<double>,
      double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> >' requested here
    auto expression = xt::make_xshared(std::move(expr));
                          ^
hmm.cpp:20:5: note: in instantiation of function template specialization 'make_shared<xt::xfunction<xt::detail::plus<double>, double,
      const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long,
      4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &> >' requested here
    make_shared(arr1+arr2); // fails
    ^
In file included from hmm.cpp:2:
In file included from /Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xarray.hpp:19:
In file included from /Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xcontainer.hpp:23:
In file included from /Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xmath.hpp:23:
In file included from /Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xaccumulator.hpp:17:
/Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xexpression.hpp:411:32: error: no member named 'data' in
      'xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long,
      4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
        XTENSOR_FORWARD_METHOD(data);
        ~~~~~~~~~~~~~~~~~~~~~~~^~~~~
/Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xexpression.hpp:324:39: note: expanded from macro 'XTENSOR_FORWARD_METHOD'
        -> decltype(std::declval<E>().name())  \
                    ~~~~~~~~~~~~~~~~~ ^
/Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xexpression.hpp:413:32: error: no member named 'data_offset' in
      'xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long,
      4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
        XTENSOR_FORWARD_METHOD(data_offset);
        ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
/Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xexpression.hpp:324:39: note: expanded from macro 'XTENSOR_FORWARD_METHOD'
        -> decltype(std::declval<E>().name())  \
                    ~~~~~~~~~~~~~~~~~ ^
/Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xexpression.hpp:415:32: error: no member named 'storage' in
      'xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long,
      4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
        XTENSOR_FORWARD_METHOD(storage);
        ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
/Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xexpression.hpp:324:39: note: expanded from macro 'XTENSOR_FORWARD_METHOD'
        -> decltype(std::declval<E>().name())  \
                    ~~~~~~~~~~~~~~~~~ ^
In file included from hmm.cpp:1:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/iostream:38:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/ios:216:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__locale:15:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string:470:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string_view:169:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__string:56:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:643:
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:2057:9: error: no matching constructor for initialization of
      'xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long,
      4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
      : __value_(_VSTD::forward<_Args>(_VSTD::get<_Indexes>(__args))...) {}
        ^        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:2159:9: note: in instantiation of function template specialization
      'std::__1::__compressed_pair_elem<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>,
      1, false>::__compressed_pair_elem<xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &> > &, 0>' requested here
        _Base2(__pc, _VSTD::move(__second_args),
        ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:3572:16: note: in instantiation of function template specialization
      'std::__1::__compressed_pair<std::__1::allocator<xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &> >, xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>
      >::__compressed_pair<std::__1::allocator<xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &> > &, xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &> > &>' requested here
            :  __data_(piecewise_construct, _VSTD::forward_as_tuple(__a),
               ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:4230:26: note: in instantiation of function template specialization
      'std::__1::__shared_ptr_emplace<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>,
      std::__1::allocator<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>
      > >::__shared_ptr_emplace<xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const
      xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &> > &>' requested here
    ::new(__hold2.get()) _CntrlBlk(__a2, _VSTD::forward<_Args>(__args)...);
                         ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:4600:29: note: in instantiation of function template specialization
      'std::__1::shared_ptr<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>
      >::make_shared<xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>
      > &>' requested here
    return shared_ptr<_Tp>::make_shared(_VSTD::forward<_Args>(__args)...);
                            ^
/Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xexpression.hpp:503:43: note: in instantiation of function template specialization
      'std::__1::make_shared<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>,
      xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>
      > &>' requested here
        return xshared_expression<E>(std::make_shared<E>(expr));
                                          ^
hmm.cpp:20:5: note: in instantiation of function template specialization 'make_shared<xt::xfunction<xt::detail::plus<double>, double,
      const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long,
      4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &> >' requested here
    make_shared(arr1+arr2); // fails
    ^
/Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xfunction.hpp:557:9: note: candidate constructor not viable: no known conversion
      from 'xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>
      >' to 'const xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>'
      for 1st argument
        xfunction(const xfunction&) = default;
        ^
/Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xfunction.hpp:560:9: note: candidate constructor not viable: no known conversion
      from 'xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>
      >' to 'xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long,
      4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>' for 1st argument
        xfunction(xfunction&&) = default;
        ^
/Users/ray_zhang/anaconda3/envs/cling/include/xtensor/xfunction.hpp:553:9: note: candidate template ignored: requirement
      '!std::is_base_of<std::decay_t<xexpression<xfunction<plus<double>, double, const xarray_container<uvector<double, allocator<double>
      >, xt::layout_type::row_major, svector<unsigned long, 4, allocator<unsigned long>, true>, xtensor_expression_tag> &, const
      xarray_container<uvector<double, allocator<double> >, xt::layout_type::row_major, svector<unsigned long, 4, allocator<unsigned
      long>, true>, xtensor_expression_tag> &> > &>, self_type>::value' was not satisfied [with Func =
      xt::xexpression<xt::xfunction<xt::detail::plus<double>, double, const xt::xarray_container<xt::uvector<double,
      std::__1::allocator<double> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, const xt::xarray_container<xt::uvector<double, std::__1::allocator<double> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &>
      > &, CTA = <>]
        xfunction(Func&& f, CTA&&... e) noexcept;
        ^
5 errors generated.
```

--- a/include/xtensor/xdynamic_view.hpp
+++ b/include/xtensor/xdynamic_view.hpp
@@ -80,6 +80,9 @@ namespace xt
 
         using iterable_base = xiterable<self_type>;
         using inner_shape_type = typename iterable_base::inner_shape_type;
+        using inner_strides_type = typename base_type::inner_strides_type;
+        using inner_backstrides_type = typename base_type::inner_backstrides_type;
+
         using shape_type = typename base_type::shape_type;
         using strides_type = typename base_type::strides_type;
         using backstrides_type = typename base_type::backstrides_type;
@@ -121,7 +124,7 @@ namespace xt
         // Explicitly deleting strides method to avoid compilers complaining
         // about not being able to call the strides method from xstrided_view_base
         // private base
-        const strides_type& strides() const = delete;
+        const inner_strides_type& strides() const noexcept = delete;
 
         reference operator()();
         const_reference operator()() const;
@@ -193,7 +196,7 @@ namespace xt
         using offset_type = typename base_type::offset_type;
 
         slice_vector_type m_slices;
-        strides_type m_adj_strides;
+        inner_strides_type m_adj_strides;
 
         container_iterator data_xbegin() noexcept;
         const_container_iterator data_xbegin() const noexcept;

--- a/include/xtensor/xexpression.hpp
+++ b/include/xtensor/xexpression.hpp
@@ -14,6 +14,7 @@
 #include <vector>
 
 #include <xtl/xclosure.hpp>
+#include <xtl/xmeta_utils.hpp>
 #include <xtl/xtype_traits.hpp>
 
 #include "xlayout.hpp"
@@ -326,6 +327,33 @@ namespace xt
         return m_ptr->name();                  \
     }
 
+    namespace detail
+    {
+        template <class E>
+        struct expr_strides_type
+        {
+            using type = typename E::strides_type;
+        };
+
+        template <class E>
+        struct expr_inner_strides_type
+        {
+            using type = typename E::inner_strides_type;
+        };
+
+        template <class E>
+        struct expr_backstrides_type
+        {
+            using type = typename E::backstrides_type;
+        };
+
+        template <class E>
+        struct expr_inner_backstrides_type
+        {
+            using type = typename E::inner_backstrides_type;
+        };
+    }
+
     /**
      * @class xshared_expression
      * @brief Shared xexpressions
@@ -367,14 +395,21 @@ namespace xt
         using size_type = typename E::size_type;
         using difference_type = typename E::difference_type;
 
+        using inner_shape_type = typename E::inner_shape_type;
         using shape_type = typename E::shape_type;
 
-        using inner_shape_type = typename E::inner_shape_type;
-
-        using strides_type = typename E::strides_type;
-        using backstrides_type = typename E::backstrides_type;
-        using inner_strides_type = typename E::inner_strides_type;
-        using inner_backstrides_type = typename E::inner_backstrides_type;
+        using strides_type = xtl::mpl::eval_if_t<has_strides<E>,
+                                                 detail::expr_strides_type<E>,
+                                                 get_strides_type<shape_type>>;
+        using backstrides_type = xtl::mpl::eval_if_t<has_strides<E>,
+                                                     detail::expr_backstrides_type<E>,
+                                                     get_strides_type<shape_type>>;
+        using inner_strides_type = xtl::mpl::eval_if_t<has_strides<E>,
+                                                       detail::expr_inner_strides_type<E>,
+                                                       get_strides_type<shape_type>>;
+        using inner_backstrides_type = xtl::mpl::eval_if_t<has_strides<E>,
+                                                           detail::expr_inner_backstrides_type<E>,
+                                                           get_strides_type<shape_type>>;
 
         using stepper = typename E::stepper;
         using const_stepper = typename E::const_stepper;
@@ -405,15 +440,55 @@ namespace xt
         XTENSOR_FORWARD_METHOD(storage_end);
         XTENSOR_FORWARD_METHOD(storage_cend);
         XTENSOR_FORWARD_METHOD(layout);
-       
-        template <class CE = E, class = std::enable_if_t<has_data_interface<CE>::value, int>>
-        XTENSOR_FORWARD_METHOD(strides);
-        template <class CE = E, class = std::enable_if_t<has_data_interface<CE>::value, int>>
-        XTENSOR_FORWARD_METHOD(data);
-        template <class CE = E, class = std::enable_if_t<has_data_interface<CE>::value, int>>
-        XTENSOR_FORWARD_METHOD(data_offset);
-        template <class CE = E, class = std::enable_if_t<has_data_interface<CE>::value, int>>
-        XTENSOR_FORWARD_METHOD(storage);
+        
+        template <class T = E>
+        std::enable_if_t<has_strides<T>::value, const inner_strides_type&>
+        strides() const
+        {
+            return m_ptr->strides();
+        }
+
+        template <class T = E>
+        std::enable_if_t<has_strides<T>::value, const inner_strides_type&>
+        backstrides() const
+        {
+            return m_ptr->backstrides();
+        }
+
+        template <class T = E>
+        std::enable_if_t<has_data_interface<T>::value, pointer>
+        data() noexcept
+        {
+            return m_ptr->data();
+        }
+
+        template <class T = E>
+        std::enable_if_t<has_data_interface<T>::value, pointer>
+        data() const noexcept
+        {
+            return m_ptr->data();
+        }
+
+        template <class T = E>
+        std::enable_if_t<has_data_interface<T>::value, size_type>
+        data_offset() const noexcept
+        {
+            return m_ptr->data_offset();
+        }
+
+        template <class T = E>
+        std::enable_if_t<has_data_interface<T>::value, typename T::storage_type&>
+        storage() noexcept
+        {
+            return m_ptr->storage();
+        }
+
+        template <class T = E>
+        std::enable_if_t<has_data_interface<T>::value, const typename T::storage_type&>
+        storage() const noexcept
+        {
+            return m_ptr->storage();
+        }
        
         template <class It>
         auto element(It first, It last) {

--- a/include/xtensor/xgenerator.hpp
+++ b/include/xtensor/xgenerator.hpp
@@ -72,7 +72,6 @@ namespace xt
         using iterable_base = xconst_iterable<self_type>;
         using inner_shape_type = typename iterable_base::inner_shape_type;
         using shape_type = inner_shape_type;
-        using strides_type = get_strides_t<S>;
 
         using stepper = typename iterable_base::stepper;
         using const_stepper = typename iterable_base::const_stepper;

--- a/include/xtensor/xindex_view.hpp
+++ b/include/xtensor/xindex_view.hpp
@@ -80,7 +80,6 @@ namespace xt
         using iterable_base = xiterable<self_type>;
         using inner_shape_type = typename iterable_base::inner_shape_type;
         using shape_type = inner_shape_type;
-        using strides_type = get_strides_t<shape_type>;
 
         using indices_type = I;
 

--- a/include/xtensor/xstrided_view_base.hpp
+++ b/include/xtensor/xstrided_view_base.hpp
@@ -45,6 +45,10 @@ namespace xt
         using strides_type = get_strides_t<shape_type>;
         using backstrides_type = strides_type;
 
+        using inner_shape_type = shape_type;
+        using inner_strides_type = strides_type;
+        using inner_backstrides_type = backstrides_type;
+
         static constexpr layout_type static_layout = L;
         static constexpr bool contiguous_layout = static_layout != layout_type::dynamic;
 
@@ -61,9 +65,9 @@ namespace xt
 
         size_type size() const noexcept;
         size_type dimension() const noexcept;
-        const shape_type& shape() const noexcept;
-        const strides_type& strides() const noexcept;
-        const backstrides_type& backstrides() const noexcept;
+        const inner_shape_type& shape() const noexcept;
+        const inner_strides_type& strides() const noexcept;
+        const inner_backstrides_type& backstrides() const noexcept;
         layout_type layout() const noexcept;
 
         reference operator()();
@@ -142,9 +146,9 @@ namespace xt
 
         CT m_e;
         inner_storage_type m_storage;
-        shape_type m_shape;
-        strides_type m_strides;
-        backstrides_type m_backstrides;
+        inner_shape_type m_shape;
+        inner_strides_type m_strides;
+        inner_backstrides_type m_backstrides;
         std::size_t m_offset;
         layout_type m_layout;
     };
@@ -387,7 +391,7 @@ namespace xt
      * Returns the shape of the xtrided_view_base.
      */
     template <class CT, class S, layout_type L, class FST>
-    inline auto xstrided_view_base<CT, S, L, FST>::shape() const noexcept -> const shape_type&
+    inline auto xstrided_view_base<CT, S, L, FST>::shape() const noexcept -> const inner_shape_type&
     {
         return m_shape;
     }
@@ -396,7 +400,7 @@ namespace xt
      * Returns the strides of the xtrided_view_base.
      */
     template <class CT, class S, layout_type L, class FST>
-    inline auto xstrided_view_base<CT, S, L, FST>::strides() const noexcept -> const strides_type&
+    inline auto xstrided_view_base<CT, S, L, FST>::strides() const noexcept -> const inner_strides_type&
     {
         return m_strides;
     }
@@ -405,7 +409,7 @@ namespace xt
      * Returns the backstrides of the xtrided_view_base.
      */
     template <class CT, class S, layout_type L, class FST>
-    inline auto xstrided_view_base<CT, S, L, FST>::backstrides() const noexcept -> const backstrides_type&
+    inline auto xstrided_view_base<CT, S, L, FST>::backstrides() const noexcept -> const inner_backstrides_type&
     {
         return m_backstrides;
     }

--- a/include/xtensor/xview.hpp
+++ b/include/xtensor/xview.hpp
@@ -122,8 +122,12 @@ namespace xt
 
         using iterable_base = xiterable<self_type>;
         using inner_shape_type = typename iterable_base::inner_shape_type;
+        using inner_strides_type = get_strides_t<inner_shape_type>;
+        using inner_backstrides_type = inner_strides_type;
+
         using shape_type = inner_shape_type;
-        using strides_type = get_strides_t<shape_type>;
+        using strides_type = inner_strides_type;
+        using backstrides_type = inner_backstrides_type;
 
         using slice_type = std::tuple<S...>;
 
@@ -233,11 +237,11 @@ namespace xt
         storage() const;
 
         template <class T = xexpression_type>
-        std::enable_if_t<has_data_interface<T>::value && detail::slices_contigous<S...>::value, const strides_type&>
+        std::enable_if_t<has_data_interface<T>::value && detail::slices_contigous<S...>::value, const inner_strides_type&>
         strides() const;
 
         template <class T = xexpression_type>
-        std::enable_if_t<has_data_interface<T>::value && detail::slices_contigous<S...>::value, const strides_type&>
+        std::enable_if_t<has_data_interface<T>::value && detail::slices_contigous<S...>::value, const inner_backstrides_type&>
         backstrides() const;
 
         template <class T = xexpression_type>
@@ -315,8 +319,8 @@ namespace xt
         CT m_e;
         slice_type m_slices;
         inner_shape_type m_shape;
-        mutable strides_type m_strides;
-        mutable strides_type m_backstrides;
+        mutable inner_strides_type m_strides;
+        mutable inner_backstrides_type m_backstrides;
         mutable bool m_strides_computed;
 
         template <class... Args>
@@ -866,7 +870,7 @@ namespace xt
     template <class CT, class... S>
     template <class T>
     inline auto xview<CT, S...>::strides() const ->
-        std::enable_if_t<has_data_interface<T>::value && detail::slices_contigous<S...>::value, const strides_type&>
+        std::enable_if_t<has_data_interface<T>::value && detail::slices_contigous<S...>::value, const inner_strides_type&>
     {
         if (!m_strides_computed)
         {
@@ -879,7 +883,7 @@ namespace xt
     template <class CT, class... S>
     template <class T>
     inline auto xview<CT, S...>::backstrides() const ->
-        std::enable_if_t<has_data_interface<T>::value && detail::slices_contigous<S...>::value, const strides_type&>
+        std::enable_if_t<has_data_interface<T>::value && detail::slices_contigous<S...>::value, const inner_backstrides_type&>
     {
         if (!m_strides_computed)
         {

--- a/test/test_xexpression.cpp
+++ b/test/test_xexpression.cpp
@@ -65,6 +65,14 @@ namespace xt
         std::stringstream buffer;
         buffer << expr1;
         EXPECT_EQ(buffer.str(), "{{  2.,   4.,   6.,   8.},\n { 10.,  12.,  14.,  16.}}");
+
+        // Compilation test
+        auto sexpr1 = make_xshared(std::move(expr1));
+        using expr_type = decltype(sexpr1);
+        using strides_type = typename expr_type::strides_type;
+        using inner_strides_type = typename expr_type::inner_strides_type;
+        using backstrides_type = typename expr_type::backstrides_type;
+        using inner_strides_tybackstrides_typepe = typename expr_type::inner_backstrides_type;
     }
 
     TEST(xexpression, shared_expr_return)

